clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1154,12 +1155,24 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
         acquire();
         try {
             log.debug("Committing offsets: {} ", offsets);
+            validateOffsets(offsets);
             coordinator.commitOffsetsAsync(new HashMap<>(offsets), callback);
         } finally {
             release();
         }
     }
 
+    private void validateOffsets(Map<TopicPartition, OffsetAndMetadata> offsets) {
+        positions(offsets.keySet());
+        for (Map.Entry<TopicPartition, OffsetAndMetadata> entry : offsets.entrySet()) {
+            OffsetAndMetadata offsetAndMetadata = entry.getValue();
+            Long position = subscriptions.position(entry.getKey());
+            if (offsetAndMetadata.offset() < 0 || offsetAndMetadata.offset() > position) {
```

Comment:
By doing seek() and commit() it makes it explicit that we want to commit a larger offset and the inconvenience is pretty small.